.kokoro/build.sh

```diff
@@ -0,0 +1,237 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -eo pipefail
+
+## Get the directory of the build script
+scriptDir=$(realpath $(dirname "${BASH_SOURCE[0]}"))
+## cd to the parent directory, i.e. the root of the git repo
+cd ${scriptDir}/..
+
+# include common functions
+source ${scriptDir}/common.sh
+
+# Print out Maven & Java version
+mvn -version
+echo ${JOB_TYPE}
+
+# attempt to install 3 times with exponential backoff (starting with 10 seconds)
+retry_with_backoff 3 10 \
+  mvn install -B -V -ntp \
+    -DskipTests=true \
+    -Dclirr.skip=true \
+    -Denforcer.skip=true \
+    -Dmaven.javadoc.skip=true \
+    -Dgcloud.download.skip=true \
+    -T 1C
+
+# if GOOGLE_APPLICATION_CREDENTIALS is specified as a relative path, prepend Kokoro root directory onto it
+if [[ ! -z "${GOOGLE_APPLICATION_CREDENTIALS}" && "${GOOGLE_APPLICATION_CREDENTIALS}" != /* ]]; then
+    export GOOGLE_APPLICATION_CREDENTIALS=$(realpath ${KOKORO_GFILE_DIR}/${GOOGLE_APPLICATION_CREDENTIALS})
+fi
+
+RETURN_CODE=0
+set +e
+
+echo "Installing the cbt CLI command."
+apt install -y google-cloud-sdk-cbt
+
+run_unit_tests() {
+    echo "***Running connector's unit tests.***"
+    CONNECTOR_MODULE="spark-bigtable_2.12"
+    mvn -pl ${CONNECTOR_MODULE} -am  \
+        test -B -ntp -Dclirr.skip=true -Denforcer.skip=true -Dcheckstyle.skip
+    return $?
+}
+
+run_bigtable_spark_tests() {
+    SPARK_VERSION=$1
+    MAVEN_PROFILES=$2
+    echo "***Running Spark-Bigtable tests for Spark ${SPARK_VERSION} and profile(s) ${MAVEN_PROFILES}.***"
+    BIGTABLE_SPARK_IT_MODULE="spark-bigtable_2.12-it"
+    mvn -pl ${BIGTABLE_SPARK_IT_MODULE} \
+        failsafe:integration-test failsafe:verify \
+        -B -ntp -Dclirr.skip=true -Denforcer.skip=true -Dcheckstyle.skip \
+        -Dspark.version=${SPARK_VERSION} \
+        -DbigtableProjectId=${BIGTABLE_PROJECT_ID} \
+        -DbigtableInstanceId=${BIGTABLE_INSTANCE_ID} \
+        -P ${MAVEN_PROFILES}
+    return $?
+}
+
+get_bigtable_spark_jar() {
+    # This makes the script independent of the connector's version and
+    # ignores the source code JAR.
+    echo $(ls spark-bigtable_2.12/target/spark-bigtable_2.12-* | grep -v sources)
+}
+
+create_table_id() {
+    TEST_TYPE=$1
+    # Generate a random id with the 'cbt-TEST_TYPE-1682536895-138fa' format.
+    echo "cbt-${TEST_TYPE}-$(date +%s)-$(echo $RANDOM | md5sum | head -c 5)"
+}
+
+# Create a table with pre-splits at random bytes. Some caveats:
+#  1. character \054 (comma) gets ignored because ',' is also the delimiter. we
+#      can probably avoid this with escaping, etc., but it's not really critical
+#      since there's nothing special about its byte value to try to manually cover.
+#  2. Repeated characters would be ignored w/o affecting correctness.
+create_table_with_random_splits() {
```

Comment:
> why does this need to happen in bash? why not a class rule in java?

isn't pre-split only available in cbt CLI? Or do you mean executing a cli command inside java code?

> Also this needs to be self healing ....  if the clean up failed in the last run, then this and all subsequent test runs will fail. Also, you will have a problem with concurrent runs (ie 2 open prs) clobbering each other

we use different table names for each run, I don't think they would conflict with each other. Also, this one's only used in the fuzz filter push down test, which is run periodically, so there's only one instance of it

> I think it would be better to have a persistent table and each test run owning its own prefix. And then configure a gc policy to clean up old rows

this is an interesting idea, but I think there are two issues:
1. Since this is used in the filter fuzz test, it's harder to create a random filter in the code while respecting the row key prefix (especially for something like long, or if we add float, etc, in the future)
2. By creating a new table each time, we're adding another dimension of randomness, which is the split points. I think it'll be harder to ensure dynamic split points, especially while maintaining row key prefixes for each test run

Reply:
1. You can pre-split using both the native bigtable api and the hbase api (createTable(TableDescriptor, byte[][] splits)
2. ah, sorry missed the random table name, but on start up you still want to delete old tables that might not have been cleaned up because the test runner got killed
3. Ok, lets leave be